File: square_free/square_free.cpp
#include <cstdint>
#include <iostream>
#include <string>
#include "prime_sieve.hpp"

using integer = uint64_t;

bool square_free(const prime_sieve& sieve, integer n) {
    if (n % 4 == 0)
        return false;
    for (integer p = 3; p * p <= n; p += 2) {
        if (sieve.is_prime(p) && n % (p * p) == 0)
            return false;
    }
    return true;
}

void print_square_free_numbers(const prime_sieve& sieve, integer from, integer to) {
    std::cout << "Square-free numbers between " << from
        << " and " << to << ":\n";
    std::string line;
    for (integer i = from; i <= to; ++i) {
        if (square_free(sieve, i)) {
            if (!line.empty())
                line += ' ';
            line += std::to_string(i);
            if (line.size() >= 80) {
                std::cout << line << '\n';
                line.clear();
            }
        }
    }
    if (!line.empty())
        std::cout << line << '\n';
}

void print_square_free_count(const prime_sieve& sieve, integer from, integer to) {
    integer count = 0;
    for (integer i = from; i <= to; ++i) {
        if (square_free(sieve, i))
            ++count;
    }
    std::cout << "Number of square-free numbers between "
        << from << " and " << to << ": " << count << '\n';
}

int main() {
    prime_sieve sieve(1000001);
    print_square_free_numbers(sieve, 1, 145);
    print_square_free_numbers(sieve, 1000000000000LL, 1000000000145LL);
    print_square_free_count(sieve, 1, 100);
    print_square_free_count(sieve, 1, 1000);
    print_square_free_count(sieve, 1, 10000);
    print_square_free_count(sieve, 1, 100000);
    print_square_free_count(sieve, 1, 1000000);
    return 0;
}

File: square_free/square_free.swift
func squareFree(number: UInt) -> Bool {
    var n = number
    if n % 4 == 0 {
        return false
    }
    var p: UInt = 3
    while p * p <= n {
        var count = 0
        while n % p == 0 {
            count += 1
            if count > 1 {
                return false
            }
            n /= p
        }
        p += 2
    }
    return true
}

func printSquareFreeNumbers(from: UInt, to: UInt) {
    print("Square-free numbers between \(from) and \(to):")
    var line = ""
    for i in from...to {
        if squareFree(number: i) {
            if line.count > 0 {
                line += " "
            }
            line += "\(i)"
            if line.count >= 80 {
                print(line)
                line = ""
            }
        }
    }
    if line.count > 0 {
        print(line)
    }
}

func printSquareFreeCount(from: UInt, to: UInt) {
    var count = 0
    for i in from...to {
        if squareFree(number: i) {
            count += 1
        }
    }
    print("Number of square-free numbers between \(from) and \(to): \(count)")
}

printSquareFreeNumbers(from: 1, to: 145)
printSquareFreeNumbers(from: 1000000000000, to: 1000000000145)

var n: UInt = 100
while n <= 1000000 {
    printSquareFreeCount(from: 1, to: n)
    n *= 10
}
